src/java/org/apache/cassandra/service/accord/async/AsyncLoader.java

```diff
@@ -0,0 +1,232 @@
+    }
+
+    @VisibleForTesting
+    Function<AccordCommand, Future<?>> loadCommandFunction(Object callback)
+    {
+        return command -> Stage.READ.submit(() -> {
+            try
+            {
+                logger.trace("Starting load of {} for {}", command.txnId(), callback);
+                AccordKeyspace.loadCommand(command);
+                logger.trace("Completed load of {} for {}", command.txnId(), callback);
+            }
+            catch (Throwable t)
+            {
+                logger.error(String.format("Exception loading %s for %s", command.txnId(), callback), t);
+                throw t;
+            }
+        });
+    }
+
+    @VisibleForTesting
+    Function<AccordCommandsForKey, Future<?>> loadCommandsPerKeyFunction(Object callback)
+    {
+        return cfk -> Stage.READ.submit(() -> {
+            try
+            {
+                logger.trace("Starting load of {} for {}", cfk.key(), callback);
+                AccordKeyspace.loadCommandsForKey(cfk);
+                logger.trace("Completed load of {} for {}", cfk.key(), callback);
+            }
+            catch (Throwable t)
+            {
+                logger.error(String.format("Exception loading %s for %s", cfk.key(), callback), t);
+                throw t;
+            }
+        });
+    }
+
+    private Future<?> referenceAndDispatchReads(AsyncContext context, Object callback)
+    {
+        List<Future<?>> futures = null;
+
+        futures = referenceAndDispatchReads(txnIds,
+                                            commandStore.commandCache(),
+                                            context.commands.items,
+                                            loadCommandFunction(callback),
+                                            futures,
+                                            callback);
+
+        futures = referenceAndDispatchReads(keys,
+                                            commandStore.commandsForKeyCache(),
+                                            context.commandsForKey.items,
+                                            loadCommandsPerKeyFunction(callback),
+                                            futures,
+                                            callback);
+
+        return futures != null ? FutureCombiner.allOf(futures) : null;
+    }
+
+    public boolean load(AsyncContext context, BiConsumer<Object, Throwable> callback)
+    {
+        logger.trace("Running load for {} with state {}: {} {}", callback, state, txnIds, keys);
+        commandStore.checkInStoreThread();
+        switch (state)
+        {
+            case INITIALIZED:
+                state = State.SETUP;
+            case SETUP:
+                // notify any pending write only groups we're loading a full instance so the pending changes aren't removed
+                txnIds.forEach(commandStore.commandCache()::lockWriteOnlyGroupIfExists);
+                keys.forEach(commandStore.commandsForKeyCache()::lockWriteOnlyGroupIfExists);
```

Comment:
Do we need to bring this fully into memory? It would be nice if most operations could invoke a couple of opaque `update` methods that don't (necessarily) require `CommandsForKey` to be materialised. Only `PreAccept` really needs to be able to compute conflicting operations.

Reply:
This would have the added advantage of simplifying `TxnOperation` (or maybe removing the need for it)

Reply:
I think the only operations that load commands for keys are Preaccept and BeginRecovery. Also, just fyi, I'm doing some refactoring of the "partial commands" in the metadata persistence branch so the minimimal data set is enforced by the compiler on the accord side

Reply:
sorry, I meant the `metadata-persistence-simulator` branch

Reply:
should probably update the PRs to point there. I've been using a separate set of development branches so my work doesn't disrupt caleb and cause a bunch of unnecessary rebase pain for him while he's working

Reply:
oops, I am mid rebase (and it isn't a fun one). oh well, too late.

Reply:
the partial command change shouldn't add too much additional pain once you're rebased onto the abstract command changes

Reply:
> I think the only operations that load commands for keys are Preaccept and BeginRecovery.

`Command`, `Commit`, `Apply`, `InformOfTxn` ... haven't investigated further but there's quite a few

Reply:
`Accept` might need it too in the partial replication patch if it's necessary for `CommandsForKey.register`. I can always look at refactoring it a bit then if we prefer.

Reply:
oh that's right, anything that can call `witness` needs them to potentially calculate executeAt. There's also the call the register, though those fields could be updated without doing a read

Reply:
Ah, that might improve then with partial replication in that case, as we no longer compute `max` except on `preaccept` and `repair`, as we expect each round to have been witnessed (except `accept`), and essentially defer processing until the earlier rounds arrive (in order to save io serialising transaction repeatedly). `register` on the other hand can happen on any path, pretty much.

Reply:
oh nice. register isn't as big a deal, since it could be made a write only method